common/configurable/src/main/java/io/helidon/common/configurable/LruCache.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.common.configurable;
+
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
+
+import io.helidon.config.Config;
+
+/**
+ * Least recently used cache.
+ * This cache has a capacity. When the capacity is reached, the oldest record is removed from the cache when a new one
+ * is added.
+ *
+ * @param <K> type of the keys of the map
+ * @param <V> type of the values of the map
+ */
+public final class LruCache<K, V> {
+    /**
+     * Default capacity of the cache: {@value}.
+     */
+    public static final int DEFAULT_CAPACITY = 10000;
+
+    private final LinkedHashMap<K, V> backingMap = new LinkedHashMap<>();
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock(true);
+    private final Lock readLock = rwLock.readLock();
+    private final Lock writeLock = rwLock.writeLock();
+
+    private final int capacity;
+
+    private LruCache(Builder<K, V> builder) {
+        this.capacity = builder.capacity;
+    }
+
+    /**
+     * Create a new builder.
+     *
+     * @param <K> key type
+     * @param <V> value type
+     * @return a new fluent API builder instance
+     */
+    public static <K, V> Builder<K, V> builder() {
+        return new Builder<>();
+    }
+
+    /**
+     * Create an instance with default configuration.
+     *
+     * @param <K> key type
+     * @param <V> value type
+     * @return a new cache instance
+     * @see #DEFAULT_CAPACITY
+     */
+    public static <K, V> LruCache<K, V> create() {
+        Builder<K, V> builder = builder();
+        return builder.build();
+    }
+
+    /**
+     * Get a value from the cache.
+     *
+     * @param key key to retrieve
+     * @return value if present or empty
+     */
+    public Optional<V> get(K key) {
+        readLock.lock();
```

Comment:
If this should work as a proper LRU, the get operation should move the requested value to the end of the map (so it becomes the "youngest")